크로스 밸리데이션 방법 중에서 K 분할 교차 검증 (K-fold cross validation)

1. 집합 X를 A, B, C로 분할합니다.
2. A를 테스트 데이터, B,C 를 훈련데이터로 사용해서 분류 정밀도 s1을 구합니다.
3. B를 테스트 데이터, A,C 를 훈련데이터로 사용해서 분류 정밀도 s2을 구합니다.
4. C를 테스트 데이터, A,B 를 훈련데이터로 사용해서 분류 정밀도 s3을 구합니다.
5. 분류 정밀도 s1, s2, s3 의 평균을 구해 최종적인 분류 정밀도를 구합니다.

### skikit-learn 의 크로스 밸리데이션 사용해보기

In [2]:
import pandas as pd
from sklearn import svm, metrics, model_selection
import random,re
from sklearn import datasets

In [5]:
iris = datasets.load_iris()
data = iris.data
label = iris.target

print(type(data))
print(type(label))

<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [6]:
clf = svm.SVC()
scores = model_selection.cross_val_score(clf, data, label, cv=5)
print('각각의 정답률 = ', scores)

각각의 정답률 =  [ 0.96666667  1.          0.96666667  0.96666667  1.        ]


### 그리드 서치

매개변수 튜닝

In [8]:
import pandas as pd
from sklearn import model_selection, svm, metrics
from sklearn.grid_search import GridSearchCV

train_csv = pd.read_csv('../data/mnist/train.csv')
test_csv = pd.read_csv('../data/mnist/t10k.csv')

train_label = train_csv.iloc[:,0]
train_data = train_csv.iloc[:,1:577]
test_label = test_csv.iloc[:,0]
test_data = test_csv.iloc[:,1:577]
print('학습 데이터의 수 =', len(train_label))

params = [
    {"C":[1,10,100,1000], "kernel":["linear"]},
    {"C":[1,10,100,1000], "kernel":["rbf"], "gamma":[0.001, 0.0001]}
]

# 그리드 서치 수행
clf = GridSearchCV( svm.SVC(), params, n_jobs=-1)
clf.fit(train_data, train_label)
print('학습기 =', clf.best_estimator_)

predict = clf.predict(test_data)
ac_score = metrics.accuracy_score(predict, test_label)
print('정답률=', ac_score)

학습 데이터의 수 = 1000
학습기 = SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)
정답률= 0.866


이전의 9.786에 비해ㅓ 0.866 으로 정답률이 개선됐습니다.
n_jobs - 병렬 계산할 프로세스 수를 지정.